packages/tenants/src/context/TenantAgentContextProvider.ts

```diff
@@ -55,13 +56,26 @@ export class TenantAgentContextProvider implements AgentContextProvider {
   }
 
   public async getAgentContextForContextCorrelationId(contextCorrelationId: string) {
+    this.logger.debug('Inside getAgentContextForContextCorrelationId')
     // It could be that the root agent context is requested, in that case we return the root agent context
     if (contextCorrelationId === this.rootAgentContext.contextCorrelationId) {
       return this.rootAgentContext
     }
 
-    // TODO: maybe we can look at not having to retrieve the tenant record if there's already a context available.
-    const tenantRecord = await this.tenantRecordService.getTenantById(this.rootAgentContext, contextCorrelationId)
+    const cache = this.rootAgentContext.dependencyManager.resolve(CacheModuleConfig).cache
+
+    this.logger.debug('Getting tenantRecord from cache')
+    let tenantRecord: TenantRecord | null = await cache.get(
+      this.rootAgentContext,
+      `contextCorrelationId-${contextCorrelationId}`
+    )
+    if (!tenantRecord) {
+      // TODO: maybe we can look at not having to retrieve the tenant record if there's already a context available.
+      this.logger.debug('TenantRecord not found in cache')
+      tenantRecord = await this.tenantRecordService.getTenantById(this.rootAgentContext, contextCorrelationId)
+      await cache.set(this.rootAgentContext, `contextCorrelationId-${contextCorrelationId}`, tenantRecord)
```

Comment:
I think we should only put the tenant record in the cache after we have run the checks for agent storage updates etc.. (so after getContextForSession, and before return the agent context)

Reply:
I appreciate your suggestion. In line with that, I modified the code.